12-offer-encoding.md

```diff
@@ -0,0 +1,870 @@
+  - if it is connected only by private channels:
+    - MUST include `offer_paths` containing one or more paths to the node from
+      publicly reachable nodes.
+  - otherwise:
+    - MAY include `offer_paths`.
+  - if it includes `offer_paths`:
+    - SHOULD ignore any invoice_request which does not use the path.
+    - MAY set `offer_issuer_id`.
+  - otherwise:
+     - MUST set `offer_issuer_id` to the node's public key to request the invoice from.
+  - if it sets `offer_issuer`:
+    - SHOULD set it to identify the issuer of the invoice clearly.
+    - if it includes a domain name:
+      - SHOULD begin it with either user@domain or domain
+      - MAY follow with a space and more text
+  - if it can supply more than one item for a single invoice:
+    - if the maximum quantity is known:
+      - MUST set that maximum in `offer_quantity_max`.
+      - MUST NOT set `offer_quantity_max` to 0.
+    - otherwise:
+      - MUST set `offer_quantity_max` to 0.
+  - otherwise:
+    - MUST NOT set `offer_quantity_max`.
+
+A reader of an offer:
+  - if the offer contains any TLV fields outside the inclusive ranges: 1 to 79 and 1000000000 to 1999999999:
+    - MUST NOT respond to the offer.
+  - if `offer_features` contains unknown _odd_ bits that are non-zero:
+    - MUST ignore the bit.
+  - if `offer_features` contains unknown _even_ bits that are non-zero:
+    - MUST NOT respond to the offer.
+    - SHOULD indicate the unknown bit to the user.
+  - if `offer_chains` is not set:
+    - if the node does not accept bitcoin invoices:
+      - MUST NOT respond to the offer
+  - otherwise: (`offer_chains` is set):
+    - if the node does not accept invoices for any of the `chains`:
+      - MUST NOT respond to the offer
+  - if `offer_amount` is set and `offer_description` is not set:
+    - MUST NOT respond to the offer.
+  - if neither `offer_issuer_id` nor `offer_paths` are set:
+    - MUST NOT respond to the offer.
+  - if `num_hops` is 0 in any `blinded_path` in `offer_paths`:
+    - MUST NOT respond to the offer.
+  - if it uses `offer_amount` to provide the user with a cost estimate:
+    - MUST take into account the currency units for `offer_amount`:
+      - `offer_currency` field if set
+      - otherwise, the minimum lightning-payable unit (e.g. milli-satoshis for
+        bitcoin).
+    - MUST warn user if amount of actual invoice differs significantly
+        from that estimate.
+  - SHOULD not respond to an offer if the current time is after
+    `offer_absolute_expiry`.
+  - if it chooses to sends an `invoice_request`, it sends an onion message:
+    - if `offer_paths` is set:
+      - MUST send the onion message via any path in `offer_paths` to the final `onion_msg_hop`.`blinded_node_id` in that path
+    - otherwise:
+      - MUST send the onion message to `offer_issuer_id`
+    - MAY send more than one `invoice_request` onion message at once.
+
+## Rationale
+
+The entire offer is reflected in the invoice_request, both for
```

Comment:
> I think we should use an identifier to reference the offer instead of sending the copied info the receiver already knows of, just imagine millions of invoice requests all coming from the same offer.

How would you check the invariants if the offer is stateless?

> Also how can I validate that the `invoice_request.offer_` is indeed the original offer I created if I'm stateless?

You can do so using the bytes included in the `offer_metadata`.